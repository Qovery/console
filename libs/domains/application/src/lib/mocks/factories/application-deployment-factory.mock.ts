import { DeploymentHistoryApplication, StateEnum } from 'qovery-typescript-axios'
import { Chance } from 'chance'

const chance = new Chance()

export const applicationDeploymentsFactoryMock = (howMany: number): DeploymentHistoryApplication[] =>
  Array.from({ length: howMany }).map((_, index) => ({
    id: index.toString(),
    created_at: new Date().toString(),
    updated_at: new Date().toString(),
    name: chance.name(),
    status: chance.pickone(Object.values([StateEnum.DEPLOYED, StateEnum.RUNNING, StateEnum.STOP_ERROR])),
  }))
