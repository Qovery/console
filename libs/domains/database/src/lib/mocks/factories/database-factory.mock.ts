import { Chance } from 'chance'
import { Database, DatabaseTypeEnum, DatabaseModeEnum, DatabaseAccessibilityEnum } from 'qovery-typescript-axios'

const chance = new Chance()

export const databaseFactoryMock = (howMany: number): Database[] =>
  Array.from({ length: howMany }).map((_, index) => ({
    id: `${index}`,
    created_at: new Date().toString(),
    updated_at: new Date().toString(),
    name: chance.name(),
    type: DatabaseTypeEnum.POSTGRESQL,
    version: '12',
    mode: DatabaseModeEnum.CONTAINER,
    accessibility: DatabaseAccessibilityEnum.PRIVATE,
    cpu: 1000,
    memory: 1024,
    storage: 1024,
    environment: {
      id: chance.guid(),
    },
    host: chance.name(),
    port: 80,
    maximum_cpu: 10,
    maximum_memory: 10,
    disk_encrypted: false,
  }))
