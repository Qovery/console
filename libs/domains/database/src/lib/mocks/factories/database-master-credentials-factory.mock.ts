import { Chance } from 'chance'
import { Credentials } from 'qovery-typescript-axios'

const chance = new Chance()

export const databaseMasterCredentialsFactoryMock = (howMany: number): Credentials[] =>
  Array.from({ length: howMany }).map((_, index) => ({
    login: chance.name(),
    password: chance.guid(),
    host: chance.name(),
    port: chance.d4(),
  }))
