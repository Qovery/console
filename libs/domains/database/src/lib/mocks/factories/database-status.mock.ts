import { Chance } from 'chance'
import { Status, ServiceDeploymentStatusEnum, StateEnum } from 'qovery-typescript-axios'

const chance = new Chance()

export const databaseStatusFactoryMock = (howMany: number): Status[] =>
  Array.from({ length: howMany }).map((_, index) => ({
    id: `${index}`,
    message: chance.sentence(),
    service_deployment_status: chance.pickone(
      Object.values([
        ServiceDeploymentStatusEnum.UP_TO_DATE,
        ServiceDeploymentStatusEnum.UP_TO_DATE,
        ServiceDeploymentStatusEnum.UP_TO_DATE,
      ])
    ),
    state: chance.pickone(Object.values([StateEnum.DEPLOYED, StateEnum.RUNNING, StateEnum.STOP_ERROR])),
  }))
