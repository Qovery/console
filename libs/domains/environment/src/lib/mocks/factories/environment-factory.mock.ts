import { Chance } from 'chance'
import {
  DeploymentHistoryStatusEnum,
  EnvironmentModeEnum,
  ServiceDeploymentStatusEnum,
  StateEnum,
  WeekdayEnum,
} from 'qovery-typescript-axios'
import { EnvironmentEntity } from '@qovery/shared/interfaces'

const chance = new Chance()

const service = {
  id: chance.integer().toString(),
  created_at: chance.date().toString(),
  updated_at: chance.date().toString(),
  name: chance.name(),
  status: chance.pickone(Object.values([StateEnum.DEPLOYED, StateEnum.RUNNING, StateEnum.STOP_ERROR])),
}

const database = {
  id: chance.integer().toString(),
  created_at: chance.date().toString(),
  updated_at: chance.date().toString(),
  name: chance.name(),
  status: chance.pickone(Object.values([DeploymentHistoryStatusEnum.SUCCESS, DeploymentHistoryStatusEnum.FAILED])),
}

const deployment = {
  id: chance.integer().toString(),
  created_at: chance.date().toString(),
  updated_at: chance.date().toString(),
  status: StateEnum.DEPLOYED,
  applications: [service, service, service],
  databases: [database, database, database],
}

const deploymentRules = {
  id: chance.guid(),
  created_at: chance.date().toString(),
  updated_at: chance.date().toString(),
  auto_deploy: true,
  auto_stop: true,
  auto_delete: true,
  auto_preview: true,
  timezone: 'UTC',
  start_time: chance.date().toString(),
  stop_time: chance.date().toString(),
  weekdays: [chance.pickone(Object.values(WeekdayEnum))],
}

export const environmentFactoryMock = (howMany: number, noStatus = false, noDeployments = false): EnvironmentEntity[] =>
  Array.from({ length: howMany }).map((_, index) => ({
    id: `${(index + 1) * Math.random()}`,
    created_at: chance.date().toString(),
    updated_at: chance.date().toString(),
    name: chance.name(),
    last_updated_by: chance.date().toString(),
    cloud_provider: {
      provider: chance.name(),
      cluster: chance.name(),
    },
    mode: chance.pickone(
      Object.values([EnvironmentModeEnum.DEVELOPMENT, EnvironmentModeEnum.PRODUCTION, EnvironmentModeEnum.STAGING])
    ),
    cluster_id: chance.guid(),
    status: !noStatus
      ? {
          id: `${index}`,
          state: chance.pickone(Object.values([StateEnum.DEPLOYED, StateEnum.RUNNING, StateEnum.STOP_ERROR])),
          message: chance.word({ length: 10 }),
          service_deployment_status: chance.pickone(Object.values([ServiceDeploymentStatusEnum.UP_TO_DATE])),
        }
      : undefined,
    deployments: !noDeployments ? [deployment, deployment, deployment] : undefined,
    deploymentRules: deploymentRules,
  }))
