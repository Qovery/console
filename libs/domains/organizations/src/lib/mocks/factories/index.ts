export * from './organizations-factory.mock'
