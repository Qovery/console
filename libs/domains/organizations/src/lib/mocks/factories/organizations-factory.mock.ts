import { OrganizationInterface } from '../../interfaces/organizations.interface'
import { OrganizationPlanType } from '../../enums/organizations.enum'
import { Chance } from 'chance'

const chance = new Chance()

export const organizationsFactoryMock = (howMany: number): OrganizationInterface[] =>
  Array.from({ length: howMany }).map((_, index) => ({
    id: `${index}`,
    created_at: new Date(),
    updated_at: new Date(),
    name: chance.name(),
    description: chance.word({ length: 10 }),
    plan: chance.pickone(Object.values(OrganizationPlanType)),
    website_url: chance.url(),
    repository: chance.name(),
    logo_url: chance.url(),
    icon_url: chance.url(),
    owner: chance.name(),
  }))
