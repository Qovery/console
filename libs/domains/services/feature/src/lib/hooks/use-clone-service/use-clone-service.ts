import { useMutation, useQueryClient } from '@tanstack/react-query'
import { mutations } from '@qovery/domains/services/data-access'
import { queries } from '@qovery/state/util-queries'

export function useCloneService() {
  const queryClient = useQueryClient()
  return useMutation(mutations.cloneService, {
    onSuccess: (result) => {
      queryClient.invalidateQueries({
        queryKey: queries.services.list(result.environment.id).queryKey,
      })
    },
    meta: {
      notifyOnSuccess: {
        title: 'Your service is cloned',
      },
      notifyOnError: true,
    },
  })
}

export default useCloneService
