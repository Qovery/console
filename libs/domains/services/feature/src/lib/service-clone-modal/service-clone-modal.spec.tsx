import { useNavigate } from 'react-router-dom'
import selectEvent from 'react-select-event'
import { environmentFactoryMock } from '@qovery/shared/factories'
import { renderWithProviders, screen } from '@qovery/shared/util-tests'
import { useCloneService } from '../hooks/use-clone-service/use-clone-service'
import ServiceCloneModal, { type ServiceCloneModalProps } from './service-clone-modal'

const mockEnvironments = environmentFactoryMock(3)

jest.mock('../hooks/use-clone-service/use-clone-service', () => {
  const mock = {
    // https://github.com/jestjs/jest/issues/10894
    mutateAsync: jest.fn(),
  }
  return {
    ...jest.requireActual('../hooks/use-clone-service/use-clone-service'),
    useCloneService: () => mock,
  }
})

jest.mock('../hooks/use-service/use-service', () => ({
  ...jest.requireActual('../hooks/use-service/use-service'),
  useService: () => ({
    data: {
      id: '1',
      serviceType: 'APPLICATION',
      name: 'my-service',
    },
  }),
}))

jest.mock('../hooks/use-environments/use-environments', () => ({
  ...jest.requireActual('../hooks/use-environments/use-environments'),
  useEnvironments: () => ({ data: mockEnvironments, isLoading: false }),
}))

jest.mock('react-router-dom', () => {
  const navigate = jest.fn()
  return {
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => navigate,
  }
})

const props: ServiceCloneModalProps = {
  onClose: jest.fn(),
  serviceId: '1',
  organizationId: '0',
  projectId: '1',
}

describe('ServiceCloneModal', () => {
  it('should match snapshot', async () => {
    const { container } = renderWithProviders(<ServiceCloneModal {...props} />)
    expect(container).toMatchSnapshot()
  })

  it('should submit form on click on button', async () => {
    // https://github.com/jestjs/jest/issues/10894
    useCloneService().mutateAsync = jest.fn().mockResolvedValue({ id: 1 })

    const { userEvent } = renderWithProviders(<ServiceCloneModal {...props} />)

    const input = screen.getByRole('textbox', { name: /new service name/i })
    await userEvent.clear(input)
    await userEvent.type(input, 'test')

    await selectEvent.select(screen.getByLabelText(/environment/i), mockEnvironments[2].name, {
      container: document.body,
    })

    const submitButton = screen.getByRole('button', { name: /clone/i })
    await userEvent.click(submitButton)

    expect(useCloneService().mutateAsync).toHaveBeenCalledWith({
      serviceId: '1',
      serviceType: 'APPLICATION',
      payload: {
        environment_id: mockEnvironments[2].id,
        name: 'test',
      },
    })
    expect(useNavigate()).toHaveBeenCalledWith(
      '/organization/0/project/1/environment/2951580907208704/application/1/general'
    )
  })
})
