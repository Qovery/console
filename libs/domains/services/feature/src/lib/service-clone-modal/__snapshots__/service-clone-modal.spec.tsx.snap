// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceCloneModal should match snapshot 1`] = `
<div>
  <div
    class="p-5"
  >
    <h2
      class="h4 max-w-sm truncate text-neutral-400 dark:text-neutral-50"
    >
      Clone Service
    </h2>
    <p
      class="mt-2 text-sm text-neutral-350 dark:text-neutral-50"
    >
      Clone the service on the same or different target environment.
    </p>
    <span
      aria-controls="radix-:r6:"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="mt-2 cursor-pointer text-sm font-medium text-brand-500 transition hover:text-brand-600"
      data-state="closed"
      type="button"
    >
      Show how it works 
      <i
        aria-hidden="true"
        class="fa-regular fa-circle-question text-xs"
      />
    </span>
    <form
      class="mt-6"
    >
      <div
        class="mb-6"
        data-testid="input-text-wrapper"
      >
        <div
          class="relative"
        >
          <div
            aria-label="input-container"
            class="input input--disabled input--disabled !border-neutral-250  input--label-up "
          >
            <div
              class="pointer-events-none"
            >
              <label
                class="translate-y-2 text-sm"
                for="Service to clone"
              >
                Service to clone
              </label>
              <input
                class="input__value "
                data-testid="input-text"
                disabled=""
                id="Service to clone"
                name="clone"
                type="text"
                value="my-service"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="mb-6"
        data-testid="input-text-wrapper"
      >
        <div
          class="relative"
        >
          <div
            aria-label="input-container"
            class="input    input--label-up "
          >
            <div
              class=""
            >
              <label
                class="translate-y-2 text-sm"
                for="New service name"
              >
                New service name
              </label>
              <input
                class="input__value "
                data-testid="input-text"
                id="New service name"
                name="name"
                type="text"
                value="my-service-clone"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="mb-6"
      >
        <div
          class="input input--select    "
          data-testid="select"
        >
          <label
            class="top-1.5 translate-y-2 text-sm"
            for="Project"
          >
            Project
          </label>
          <div
            class=" css-b62m3t-container"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-2-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
            />
            <div
              class="input-select__control css-13cymwt-control"
            >
              <div
                class="input-select__value-container css-1fdsijx-ValueContainer"
              >
                <div
                  class="input-select__placeholder css-1jqq78o-placeholder"
                  id="react-select-2-placeholder"
                >
                  Select...
                </div>
                <div
                  class="input-select__input-container css-qbdosj-Input"
                  data-value=""
                >
                  <input
                    aria-autocomplete="list"
                    aria-describedby="react-select-2-placeholder"
                    aria-expanded="false"
                    aria-haspopup="true"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    class="input-select__input"
                    id="Project"
                    role="combobox"
                    spellcheck="false"
                    style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="input-select__indicators css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class="input-select__indicator-separator css-1u9des2-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class="input-select__indicator input-select__dropdown-indicator css-1xc3v61-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <input
              name="Project"
              type="hidden"
              value=""
            />
          </div>
          <input
            name="Project"
            type="hidden"
            value=""
          />
          <div
            class="pointer-events-none absolute right-4 top-1/2 -translate-y-1/2"
          >
            <span
              class="icon-solid-angle-down shrink-0 text-sm text-neutral-400"
              role="img"
            />
          </div>
        </div>
      </div>
      <div
        class="mb-6"
      >
        <div
          class="input input--select    "
          data-testid="select"
        >
          <label
            class="top-1.5 translate-y-2 text-sm"
            for="Environment"
          >
            Environment
          </label>
          <div
            class=" css-b62m3t-container"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-3-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
            />
            <div
              class="input-select__control css-13cymwt-control"
            >
              <div
                class="input-select__value-container css-1fdsijx-ValueContainer"
              >
                <div
                  class="input-select__placeholder css-1jqq78o-placeholder"
                  id="react-select-3-placeholder"
                >
                  Select...
                </div>
                <div
                  class="input-select__input-container css-qbdosj-Input"
                  data-value=""
                >
                  <input
                    aria-autocomplete="list"
                    aria-describedby="react-select-3-placeholder"
                    aria-expanded="false"
                    aria-haspopup="true"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    class="input-select__input"
                    id="Environment"
                    role="combobox"
                    spellcheck="false"
                    style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="input-select__indicators css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class="input-select__indicator-separator css-1u9des2-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class="input-select__indicator input-select__dropdown-indicator css-1xc3v61-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <input
              name="Environment"
              type="hidden"
              value=""
            />
          </div>
          <input
            name="Environment"
            type="hidden"
            value=""
          />
          <div
            class="pointer-events-none absolute right-4 top-1/2 -translate-y-1/2"
          >
            <span
              class="icon-solid-angle-down shrink-0 text-sm text-neutral-400"
              role="img"
            />
          </div>
        </div>
      </div>
      <div
        class="mt-6 flex justify-end gap-3"
      >
        <button
          class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 hover:[&:not(:focus-visible):not(:disabled)]:shadow-sm active:[&:not(:focus-visible)]:shadow-none outline-brand-500 text-sm h-11 px-5 rounded border border-transparent text-neutral-350 hover:[&:not(:active):not(:focus-visible)]:border-neutral-300 hover:[&:not(:disabled)]:text-neutral-400 dark:hover:[&:not(:disabled)]:text-neutral-250 focus-visible:text-neutral-400 active:bg-neutral-150 dark:active:bg-transparent active:text-neutral-400 disabled:text-neutral-300 disabled:bg-neutral-150 disabled:border-none"
          data-testid="cancel-button"
          type="button"
        >
          Cancel
        </button>
        <button
          class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 shadow-sm disabled:shadow-none active:shadow-none outline-neutral-500 text-sm h-11 px-5 rounded bg-brand-500 active:bg-brand-600 hover:bg-brand-400 text-white disabled:text-brand-300 disabled:bg-brand-100"
          data-testid="submit-button"
          disabled=""
          type="submit"
        >
          Clone
        </button>
      </div>
    </form>
  </div>
</div>
`;
